src/levanter/data/shardcache.py

```diff
@@ -0,0 +1,712 @@
+# Dataset for preprocessing data, tokenizing, and caching to disk.
+import asyncio
+import logging
+import sys
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from typing import IO, Dict, Generic, Iterator, List, Optional, Protocol, Sequence, Tuple, TypeVar
+
+import fsspec.core
+import pyarrow as pa
+import pyarrow.parquet as pq
+import ray
+import tblib
+from dataclasses_json import dataclass_json
+from tqdm import tqdm
+
+from levanter.data import Dataset
+
+
+# We want to support the following:
+#  1) Deterministic ordering over the data, even for a changing number of readers or writers
```

Comment:
General feedback regarding the documentation and the design doc is that there is some ambiguous verbiage; it would be useful to have a definitions section for the terms, and then use those terms consistently. In particular:

- "Input" referring to the input to the preprocessors versus to the model trainer (I suppose the term "preprocess" could also be ambiguous because the model could do its own preprocessing...)
- "Readers" could refer to the preprocessors or the model trainer
- "Resumption" could refer to the trainer or the preprocessor (since they are now decoupled; the preprocessor can write chunks faster than the trainer can read them)
- "Shards" (which refers to the input to the preprocessor) and "chunks" (which refer to the input to the trainer)